WIP: add `minio` test script for local development

ericphanson opened the pull request:
cc @omus who suggested that this could be helpful.

This almost works locally:
```julia
julia --project
;cd                _
   _       _ _(_)_     |  Documentation: https://docs.julialang.org
  (_)     | (_) (_)    |
   _ _   _| |_  __ _   |  Type "?" for help, "]?" for Pkg help.
  | | | | | | |/ _` |  |
  | | |_| | | | (_| |  |  Version 1.6.1 (2021-04-23)
 _/ |\__'_|_|_|\__'_|  |  Official https://julialang.org/ release
|__/                   |

shell> cd test
/Users/eph/AWSS3.jl/test

julia> include("minio.jl")
[ Info: Precompiling AWSS3 [1c724243-ef5b-51ab-93f4-b0a88ac62a95]
[ Info: Initializing PathSet{S3Path{MinioConfig}}
s3_folders_and_files: Error During Test at /Users/eph/AWSS3.jl/test/s3path.jl:150
  Got exception outside of a @test
  AWSException: XMinioParentIsObject -- Object-prefix is already an object, please choose a different object-prefix name.
  HTTP.ExceptionRequest.StatusError(400, "PUT", "/ocaws.jl.test.20210618t163010z/pathset-root/foobar/car.txt", HTTP.Messages.Response:
  """
  HTTP/1.1 400 Bad Request
  Accept-Ranges: bytes
  Content-Length: 452
  Content-Security-Policy: block-all-mixed-content
  Content-Type: application/xml
  Server: MinIO
  Vary: Origin
  X-Amz-Request-Id: 1689BA2A3591C6F8
  X-Xss-Protection: 1; mode=block
  Date: Fri, 18 Jun 2021 16:30:31 GMT
  
  <?xml version="1.0" encoding="UTF-8"?>
  <Error><Code>XMinioParentIsObject</Code><Message>Object-prefix is already an object, please choose a different object-prefix name.</Message><Key>pathset-root/foobar/car.txt</Key><BucketName>ocaws.jl.test.20210618t163010z</BucketName><Resource>/ocaws.jl.test.20210618t163010z/pathset-root/foobar/car.txt</Resource><RequestId>1689BA2A3591C6F8</RequestId><HostId>fbc72e52-c23e-4060-88e8-eed760c5149c</HostId></Error>""")
  
  Stacktrace:
    [1] request(::Type{ExceptionLayer{ConnectionPoolLayer{StreamLayer{Union{}}}}}, ::URIs.URI, ::Vararg{Any, N} where N; kw::Base.Iterators.Pairs{Symbol, Union{Nothing, Bool}, Tuple{Symbol, Symbol, Symbol}, NamedTuple{(:iofunction, :require_ssl_verification, :response_stream), Tuple{Nothing, Bool, Nothing}}})
      @ HTTP.ExceptionRequest ~/.julia/packages/HTTP/CGviq/src/ExceptionRequest.jl:22
    [2] request(::Type{MessageLayer{ExceptionLayer{ConnectionPoolLayer{StreamLayer{Union{}}}}}}, method::String, url::URIs.URI, headers::Vector{Pair{SubString{String}, SubString{String}}}, body::Vector{UInt8}; http_version::VersionNumber, target::String, parent::Nothing, iofunction::Nothing, kw::Base.Iterators.Pairs{Symbol, Union{Nothing, Bool}, Tuple{Symbol, Symbol}, NamedTuple{(:require_ssl_verification, :response_stream), Tuple{Bool, Nothing}}})
      @ HTTP.MessageRequest ~/.julia/packages/HTTP/CGviq/src/MessageRequest.jl:66
    [3] request(::Type{BasicAuthLayer{MessageLayer{ExceptionLayer{ConnectionPoolLayer{StreamLayer{Union{}}}}}}}, method::String, url::URIs.URI, headers::Vector{Pair{SubString{String}, SubString{String}}}, body::Vector{UInt8}; kw::Base.Iterators.Pairs{Symbol, Union{Nothing, Bool}, Tuple{Symbol, Symbol}, NamedTuple{(:require_ssl_verification, :response_stream), Tuple{Bool, Nothing}}})
      @ HTTP.BasicAuthRequest ~/.julia/packages/HTTP/CGviq/src/BasicAuthRequest.jl:28
    [4] macro expansion
      @ ~/.julia/packages/AWS/cUhwq/src/utilities/request.jl:147 [inlined]
    [5] macro expansion
      @ ~/.julia/packages/Retry/vS1bg/src/repeat_try.jl:192 [inlined]
    [6] _http_request(request::Request)
      @ AWS ~/.julia/packages/AWS/cUhwq/src/utilities/request.jl:140
    [7] macro expansion
      @ ~/.julia/packages/Mocking/U41JO/src/mock.jl:29 [inlined]
    [8] macro expansion
      @ ~/.julia/packages/AWS/cUhwq/src/utilities/request.jl:57 [inlined]
    [9] macro expansion
      @ ~/.julia/packages/Retry/vS1bg/src/repeat_try.jl:192 [inlined]
   [10] submit_request(aws::MinioConfig, request::Request; return_headers::Bool)
      @ AWS ~/.julia/packages/AWS/cUhwq/src/utilities/request.jl:54
   [11] (::RestXMLService)(request_method::String, request_uri::String, args::Dict{String, Any}; aws_config::MinioConfig)
      @ AWS ~/.julia/packages/AWS/cUhwq/src/AWS.jl:229
   [12] #put_object#170
      @ ~/.julia/packages/AWS/cUhwq/src/services/s3.jl:3451 [inlined]
   [13] s3_put(aws::MinioConfig, bucket::SubString{String}, path::String, data::Vector{UInt8}, data_type::String, encoding::String; acl::String, metadata::Dict{String, String}, tags::Dict{String, String}, kwargs::Base.Iterators.Pairs{Union{}, Union{}, Tuple{}, NamedTuple{(), Tuple{}}})
      @ AWSS3 ~/AWSS3.jl/src/AWSS3.jl:592
   [14] s3_put (repeats 2 times)
      @ ~/AWSS3.jl/src/AWSS3.jl:551 [inlined]
   [15] write(fp::S3Path{MinioConfig}, content::Vector{UInt8}; part_size_mb::Int64, multipart::Bool, other_kwargs::Base.Iterators.Pairs{Union{}, Union{}, Tuple{}, NamedTuple{(), Tuple{}}})
      @ AWSS3 ~/AWSS3.jl/src/s3path.jl:366
   [16] write
      @ ~/AWSS3.jl/src/s3path.jl:364 [inlined]
   [17] #write#81
      @ ~/AWSS3.jl/src/s3path.jl:360 [inlined]
   [18] write
      @ ~/AWSS3.jl/src/s3path.jl:360 [inlined]
   [19] macro expansion
      @ ~/AWSS3.jl/test/s3path.jl:164 [inlined]
   [20] macro expansion
      @ /Users/julia/buildbot/worker/package_macos64/build/usr/share/julia/stdlib/v1.6/Test/src/Test.jl:1151 [inlined]
   [21] test_s3_folders_and_files(ps::PathSet{S3Path{MinioConfig}})
      @ Main ~/AWSS3.jl/test/s3path.jl:152
   [22] test(ps::PathSet{S3Path{MinioConfig}}, test_sets::Vector{Function})
      @ FilePathsBase.TestPaths ~/.julia/packages/FilePathsBase/YFK4h/src/test.jl:967
   [23] macro expansion
      @ ~/AWSS3.jl/test/s3path.jl:245 [inlined]
   [24] macro expansion
      @ /Users/julia/buildbot/worker/package_macos64/build/usr/share/julia/stdlib/v1.6/Test/src/Test.jl:1151 [inlined]
   [25] top-level scope
      @ ~/AWSS3.jl/test/s3path.jl:183
   [26] include(fname::String)
      @ Base.MainInclude ./client.jl:444
   [27] macro expansion
      @ ~/AWSS3.jl/test/minio.jl:44 [inlined]
   [28] macro expansion
      @ /Users/julia/buildbot/worker/package_macos64/build/usr/share/julia/stdlib/v1.6/Test/src/Test.jl:1151 [inlined]
   [29] top-level scope
      @ ~/AWSS3.jl/test/minio.jl:44
   [30] include(fname::String)
      @ Base.MainInclude ./client.jl:444
   [31] top-level scope
      @ REPL[2]:1
   [32] eval
      @ ./boot.jl:360 [inlined]
   [33] eval_user_input(ast::Any, backend::REPL.REPLBackend)
      @ REPL /Users/julia/buildbot/worker/package_macos64/build/usr/share/julia/stdlib/v1.6/REPL/src/REPL.jl:139
   [34] repl_backend_loop(backend::REPL.REPLBackend)
      @ REPL /Users/julia/buildbot/worker/package_macos64/build/usr/share/julia/stdlib/v1.6/REPL/src/REPL.jl:200
   [35] start_repl_backend(backend::REPL.REPLBackend, consumer::Any)
      @ REPL /Users/julia/buildbot/worker/package_macos64/build/usr/share/julia/stdlib/v1.6/REPL/src/REPL.jl:185
   [36] run_repl(repl::REPL.AbstractREPL, consumer::Any; backend_on_current_task::Bool)
      @ REPL /Users/julia/buildbot/worker/package_macos64/build/usr/share/julia/stdlib/v1.6/REPL/src/REPL.jl:317
   [37] run_repl(repl::REPL.AbstractREPL, consumer::Any)
      @ REPL /Users/julia/buildbot/worker/package_macos64/build/usr/share/julia/stdlib/v1.6/REPL/src/REPL.jl:305
   [38] (::Base.var"#874#876"{Bool, Bool, Bool})(REPL::Module)
      @ Base ./client.jl:387
   [39] #invokelatest#2
      @ ./essentials.jl:708 [inlined]
   [40] invokelatest
      @ ./essentials.jl:706 [inlined]
   [41] run_main_repl(interactive::Bool, quiet::Bool, banner::Bool, history_file::Bool, color_set::Bool)
      @ Base ./client.jl:372
   [42] exec_options(opts::Base.JLOptions)
      @ Base ./client.jl:302
   [43] _start()
      @ Base ./client.jl:485
AWSS3.jl: Error During Test at /Users/eph/AWSS3.jl/test/minio.jl:43
  Got exception outside of a @test
  LoadError: AWSException: SignatureDoesNotMatch -- The request signature we calculated does not match the signature you provided. Check your key and signing method.
  HTTP.ExceptionRequest.StatusError(403, "GET", "/ocaws.jl.test.20210618t163010z?versions&versions=&prefix=", HTTP.Messages.Response:
  """
  HTTP/1.1 403 Forbidden
  Accept-Ranges: bytes
  Content-Length: 419
  Content-Security-Policy: block-all-mixed-content
  Content-Type: application/xml
  Server: MinIO
  Vary: Origin
  X-Amz-Request-Id: 1689BA2AC2D2A15C
  X-Xss-Protection: 1; mode=block
  Date: Fri, 18 Jun 2021 16:30:33 GMT
  
  <?xml version="1.0" encoding="UTF-8"?>
  <Error><Code>SignatureDoesNotMatch</Code><Message>The request signature we calculated does not match the signature you provided. Check your key and signing method.</Message><BucketName>ocaws.jl.test.20210618t163010z</BucketName><Resource>/ocaws.jl.test.20210618t163010z</Resource><RequestId>1689BA2AC2D2A15C</RequestId><HostId>fbc72e52-c23e-4060-88e8-eed760c5149c</HostId></Error>""")
  
  Stacktrace:
    [1] request(::Type{ExceptionLayer{ConnectionPoolLayer{StreamLayer{Union{}}}}}, ::URIs.URI, ::Vararg{Any, N} where N; kw::Base.Iterators.Pairs{Symbol, Union{Nothing, Bool}, Tuple{Symbol, Symbol, Symbol}, NamedTuple{(:iofunction, :require_ssl_verification, :response_stream), Tuple{Nothing, Bool, Nothing}}})
      @ HTTP.ExceptionRequest ~/.julia/packages/HTTP/CGviq/src/ExceptionRequest.jl:22
    [2] request(::Type{MessageLayer{ExceptionLayer{ConnectionPoolLayer{StreamLayer{Union{}}}}}}, method::String, url::URIs.URI, headers::Vector{Pair{SubString{String}, SubString{String}}}, body::String; http_version::VersionNumber, target::String, parent::Nothing, iofunction::Nothing, kw::Base.Iterators.Pairs{Symbol, Union{Nothing, Bool}, Tuple{Symbol, Symbol}, NamedTuple{(:require_ssl_verification, :response_stream), Tuple{Bool, Nothing}}})
      @ HTTP.MessageRequest ~/.julia/packages/HTTP/CGviq/src/MessageRequest.jl:66
    [3] request(::Type{BasicAuthLayer{MessageLayer{ExceptionLayer{ConnectionPoolLayer{StreamLayer{Union{}}}}}}}, method::String, url::URIs.URI, headers::Vector{Pair{SubString{String}, SubString{String}}}, body::String; kw::Base.Iterators.Pairs{Symbol, Union{Nothing, Bool}, Tuple{Symbol, Symbol}, NamedTuple{(:require_ssl_verification, :response_stream), Tuple{Bool, Nothing}}})
      @ HTTP.BasicAuthRequest ~/.julia/packages/HTTP/CGviq/src/BasicAuthRequest.jl:28
    [4] macro expansion
      @ ~/.julia/packages/AWS/cUhwq/src/utilities/request.jl:147 [inlined]
    [5] macro expansion
      @ ~/.julia/packages/Retry/vS1bg/src/repeat_try.jl:192 [inlined]
    [6] _http_request(request::Request)
      @ AWS ~/.julia/packages/AWS/cUhwq/src/utilities/request.jl:140
    [7] macro expansion
      @ ~/.julia/packages/Mocking/U41JO/src/mock.jl:29 [inlined]
    [8] macro expansion
      @ ~/.julia/packages/AWS/cUhwq/src/utilities/request.jl:57 [inlined]
    [9] macro expansion
      @ ~/.julia/packages/Retry/vS1bg/src/repeat_try.jl:192 [inlined]
   [10] submit_request(aws::MinioConfig, request::Request; return_headers::Bool)
      @ AWS ~/.julia/packages/AWS/cUhwq/src/utilities/request.jl:54
   [11] (::RestXMLService)(request_method::String, request_uri::String, args::Dict{String, Any}; aws_config::MinioConfig)
      @ AWS ~/.julia/packages/AWS/cUhwq/src/AWS.jl:229
   [12] #list_object_versions#122
      @ ~/.julia/packages/AWS/cUhwq/src/services/s3.jl:2322 [inlined]
   [13] s3_list_versions(aws::MinioConfig, bucket::String, path_prefix::String; kwargs::Base.Iterators.Pairs{Union{}, Union{}, Tuple{}, NamedTuple{(), Tuple{}}})
      @ AWSS3 ~/AWSS3.jl/src/AWSS3.jl:490
   [14] s3_list_versions (repeats 2 times)
      @ ~/AWSS3.jl/src/AWSS3.jl:479 [inlined]
   [15] s3_nuke_bucket(aws::MinioConfig, bucket_name::String)
      @ AWSS3 ~/AWSS3.jl/src/AWSS3.jl:834
   [16] top-level scope
      @ ~/AWSS3.jl/test/s3path.jl:337
   [17] include(fname::String)
      @ Base.MainInclude ./client.jl:444
   [18] macro expansion
      @ ~/AWSS3.jl/test/minio.jl:44 [inlined]
   [19] macro expansion
      @ /Users/julia/buildbot/worker/package_macos64/build/usr/share/julia/stdlib/v1.6/Test/src/Test.jl:1151 [inlined]
   [20] top-level scope
      @ ~/AWSS3.jl/test/minio.jl:44
   [21] include(fname::String)
      @ Base.MainInclude ./client.jl:444
   [22] top-level scope
      @ REPL[2]:1
   [23] eval
      @ ./boot.jl:360 [inlined]
   [24] eval_user_input(ast::Any, backend::REPL.REPLBackend)
      @ REPL /Users/julia/buildbot/worker/package_macos64/build/usr/share/julia/stdlib/v1.6/REPL/src/REPL.jl:139
   [25] repl_backend_loop(backend::REPL.REPLBackend)
      @ REPL /Users/julia/buildbot/worker/package_macos64/build/usr/share/julia/stdlib/v1.6/REPL/src/REPL.jl:200
   [26] start_repl_backend(backend::REPL.REPLBackend, consumer::Any)
      @ REPL /Users/julia/buildbot/worker/package_macos64/build/usr/share/julia/stdlib/v1.6/REPL/src/REPL.jl:185
   [27] run_repl(repl::REPL.AbstractREPL, consumer::Any; backend_on_current_task::Bool)
      @ REPL /Users/julia/buildbot/worker/package_macos64/build/usr/share/julia/stdlib/v1.6/REPL/src/REPL.jl:317
   [28] run_repl(repl::REPL.AbstractREPL, consumer::Any)
      @ REPL /Users/julia/buildbot/worker/package_macos64/build/usr/share/julia/stdlib/v1.6/REPL/src/REPL.jl:305
   [29] (::Base.var"#874#876"{Bool, Bool, Bool})(REPL::Module)
      @ Base ./client.jl:387
   [30] #invokelatest#2
      @ ./essentials.jl:708 [inlined]
   [31] invokelatest
      @ ./essentials.jl:706 [inlined]
   [32] run_main_repl(interactive::Bool, quiet::Bool, banner::Bool, history_file::Bool, color_set::Bool)
      @ Base ./client.jl:372
   [33] exec_options(opts::Base.JLOptions)
      @ Base ./client.jl:302
   [34] _start()
      @ Base ./client.jl:485
  in expression starting at /Users/eph/AWSS3.jl/test/s3path.jl:337
Test Summary:                  | Pass  Error  Broken  Total
AWSS3.jl                       |  174      2       3    179
  S3Path{MinioConfig}          |  157      1       2    160
    Path constructor           |    3                     3
    show                       |    1              1      2
    parsing                    |    2                     2
    convert                    |    2                     2
    components                 |    4                     4
    parents                    |    7                     7
    descendants and ascendants |   16                    16
    join                       |    3                     3
    splitext                   |    3                     3
    basename                   |    3                     3
    filename                   |    3                     3
    extensions                 |    6                     6
    isempty                    |    2                     2
    norm                       |    4                     4
    relative                   |    1                     1
    absolute                   |    2                     2
    isdir                      |    2                     2
    isfile                     |    2                     2
    stat                       |    5                     5
    filesize                   |    1                     1
    modified                   |    2                     2
    created                    |    2                     2
    readpath                   |    4                     4
    walkpath                   |    3                     3
    read                       |    2                     2
    large write/read           |    1                     1
    write                      |    5                     5
    mkdir                      |    2                     2
    cp                         |    2                     2
    mv                         |    2                     2
    sync                       |    9                     9
    touch                      |    2                     2
    tmpname                    |    3                     3
    tmpname                    |    3                     3
    mktmp                      |    3                     3
    mktmpdir                   |    2                     2
    download                   |    3                     3
    isexecutable               |    1                     1
    isreadable                 |    4                     4
    iswritable                 |    4                     4
    s3_properties              |    4                     4
    s3_folders_and_files       |    1      1       1      3
  readdir                      |   17              1     18
ERROR: LoadError: Some tests did not pass: 174 passed, 0 failed, 2 errored, 3 broken.
in expression starting at /Users/eph/AWSS3.jl/test/minio.jl:43
```

Co-authored by @hannahilea

ericphanson commented:
I just realized AWS.jl has a `test/minio.jl` script and a GHA to install the deps, so that could be re-used here.

If we can't figure out how to get all the tests to run under Minio, users won't be able to run the full test suite locally, but perhaps that's OK. Our goal here is to make it easier to write tests, and I think that can be accomplished even if the full test suite can't be run, but rather just some subset. We could have a group of tests that are run jointly with a real bucket and minio in CI (to be sure they stay runnable with minio), and then developers can run them just with minio locally when developing. And perhaps the minio CI run wouldn't need to be gated behind bors, allowing devs to get quick feedback on there PRs.

ExpandingMan commented:
My package [Minio.jl](https://gitlab.com/ExpandingMan/Minio.jl) may also be useful here, which is basically just some simple tools for making this stuff easy.

ericphanson commented:
bors try

ericphanson commented:
bors try

ericphanson commented:
bors try

ericphanson commented:
bors cancel

ericphanson commented:
bors try

omus commented on .github/workflows/CI.yml:
```diff
@@ -50,6 +50,13 @@ jobs:
       - uses: julia-actions/julia-buildpkg@latest
+      - name: "Compat fix for Julia < v1.5"
+        if: ${{ (matrix.version == '1.3') || (matrix.version == '1.0') }}
+        run: |
+          # Remove Minio from the test project
+          write("Project.toml", read(`grep -v '^Minio =' Project.toml`, String))
+          run(`sed -i -e 's/, "Minio"//' Project.toml`)
+        shell: julia --project=. --startup=no --color=yes {0}
```
Why not use `Pkg.rm` for this?

ericphanson replied:
because it's in the test deps instead of the regular deps

omus replied:
*sigh* Pkg needs to fix this one day

ericphanson commented:
Tests pass! ready to review

omus commented on test/runtests.jl:
```diff
@@ -11,8 +11,37 @@ using FilePathsBase.TestPaths
+            include("s3path.jl")
+            include("awss3.jl")
+        end
+    catch
+        # Make sure we kill the server even if a test failed.
+        kill(minio_server)
+        rethrow()
+    end
```
```suggestion
    finally
        # Make sure we kill the server even if a test failed.
        kill(minio_server)
    end
```

omus commented on test/runtests.jl:
```diff
@@ -11,8 +11,37 @@ using FilePathsBase.TestPaths
+    end
+end
+
+aws = global_aws_config(AWSConfig(; region = "us-east-1"))
```
Do we need to specify the region?
```suggestion
aws = global_aws_config()
```

omus commented on test/runtests.jl:
```diff
@@ -11,8 +11,36 @@ using FilePathsBase.TestPaths
+    # We then run all tests with s3 directly.
+
+    # We use multiple directories so that Minio can support versioning.
+    dirs = [mktempdir() for _ =1:12]
```
Seems nicer to use a single tempdir with subdirectories if you ever need to do some manual debugging:
```suggestion
    root = mktempdir()
    dirs = [mkdir(joinpath(root, string(i))) for i in 1:12]
```

omus commented on test/runtests.jl:
```diff
@@ -11,8 +11,36 @@ using FilePathsBase.TestPaths
+    end
+end
+
+aws = global_aws_config()
```
My bad. This will just read the current config and not overwrite the `MinioConfig` when set:
```suggestion
# Use the `AWSConfig` for the following tests
aws = global_aws_config(AWSConfig())
```

omus replied:
`config = ` or `cfg = ` may be a better more general name. I understand not wanting to change it though

ericphanson replied:
ah right. Yeah, it's used extensively in the tests. We could find and replace it though.

omus commented on test/runtests.jl:
```diff
@@ -11,8 +11,36 @@ using FilePathsBase.TestPaths
+    minio_server = Minio.Server(dirs; address="localhost:$port")
+
+    # We use this boolean flag to skip some tests under Minio
+    minio = true
```
Remind me why we have this global instead of checking the `global_aws_config` type?

ericphanson replied:
just because we use it in lots of places. We could do `!(aws isa AWSConfig)` instead in these local places. (We cannot do `aws isa MinioConfig` since Minio may not be loaded)

omus replied:
Ah, good point about Minio not being loaded. I'd probably use a function to encapsulate any logic needed there. Something you could do is define `abstract type MinioConfig end` in cases where Minio.jl isn't loaded

omus commented on .github/workflows/CI.yml:
```diff
@@ -50,6 +50,13 @@ jobs:
       - uses: julia-actions/julia-buildpkg@latest
+      - name: "Compat fix for Julia < v1.5"
+        if: ${{ (matrix.version == '1.3') || (matrix.version == '1.0') }}
+        run: |
+          # Remove Minio from the test project
+          write("Project.toml", read(`grep -v '^Minio =' Project.toml`, String))
+          run(`sed -i -e 's/, "Minio"//' Project.toml`)
+        shell: julia --project=. --startup=no --color=yes {0}
```
I just realized the scope of the madness going on here.
```suggestion
        run: |
          # Remove Minio from the test project
          sed -i -e '/^Minio =/d; s/"Minio"\(, \)\{0,1\}//' Project.toml
```
Using `sed` for everything here and cutting out the Julia middle man. One small tweak I did is optionally cut out the comma in case the order is changed

ericphanson replied:
Sounds good 😄

omus replied:
Round two. The case I originally coded for would handle if "Minio" was in any position in the array except for the last element. This corrects that and also supports "Minio" being the only element in the array (which should never be the case)
```suggestion
        run: |
          # Remove Minio from the test project
          sed -i -e '/^Minio =/d; s/\(, \)"Minio"//; s/"Minio"\(, \)\{0,1\}//' Project.toml
```

ericphanson commented:
@omus thinking about some of your review comments, I pushed https://github.com/JuliaCloud/AWSS3.jl/pull/163/commits/47c6a4b0b6a963f55639bd005b2f987b4e1c6c23 which is my attempt at doing this the "right" way instead of the minimally-invasive way that I had been going for.

Instead of re-including the same files twice (which feels a bit wrong), I refactored the tests into functions that accept a config argument, and then call those functions twice, once with a MinioConfig and once with an AWSConfig. I also renamed the config to `config` instead of `aws`.

I'm not actually sure this will work without #175 if `@p_str` creates the S3Paths at macro expansion time instead of function-call time but we can give it a try:

bors try

ericphanson commented:
looks like db41c2b35ff5ac9e3a7dd671908e89e78654819a wasn't quite right so I reverted it for now.

bors try

ericphanson commented:
bors try

ericphanson commented:
bors try

omus reviewed (commented):
Ugh, I thought I submitted these review comments earlier

ericphanson commented on test/runtests.jl:
```diff
@@ -11,9 +11,42 @@ using FilePathsBase.TestPaths
+
+    try
+        run(minio_server, wait=false)
+        config = global_aws_config(MinioConfig("http://localhost:$port"; username="minioadmin", password="minioadmin"))
```
let's try it

ericphanson commented:
bors try

omus commented on test/runtests.jl:
```diff
@@ -11,9 +11,42 @@ using FilePathsBase.TestPaths
+
+# Set `AWSConfig` as the default for the following tests
+aws = global_aws_config(AWSConfig())
 @testset "AWSS3.jl" begin
```
I think all of the `@testset`s and `@test`s should be nested under "AWSS3.jl" and this should be renamed to "S3".

ericphanson replied:
done!

omus reviewed (approved):
LGTM otherwise. @mattBrzezinski could you also review this?

ericphanson commented:
bors try

mattBrzezinski reviewed (approved):
These changes LGTM. Only comment I have is I can't wait for the new LTS. :'(

ericphanson commented:
I'm so confused by the test failure: https://github.com/JuliaCloud/AWSS3.jl/runs/3312685592#step:7:165. Could it be some kinda weird network issue?

I'll try again just in case:

bors try
